### ISSUE TITLE
model_rebuild is surprisingly slow, can we make it lazyloaded

### ISSUE THREAD

**Issue by ghost** [mirumee/ariadne-codegen#226]
We found that it takes more than 8 seconds to import the generated Client because we have more than 1k graphql models so that `model_rebuild` is called more than 1k times, which is super slow

```bash
python -X importtime -c "from graphql import Client"

....
import time:      1321 |       1321 |       pydantic._internal._std_types_schema
import time:      4473 |       5794 |     graphql.prompts_prompts_by_pk
import time:       198 |       5991 |   graphql.client
import time:      8467 |       8467 |   graphql.enums
import time:   7874752 |    7874752 |   graphql.input_types
import time:       760 |    8139339 | graphql
```

Is it possible to make `model_rebuild` lazyloaded?

**Comment by mat-sop** [comment]
Hey, I haven't looked in-depth into the problem yet, but for now, you can reasonably easily automate your solution by using plugin hooks, e.g.

Separate `LazyInitMeta` and `CombinedMeta` to `meta.py`:
```py
from .base_model import BaseModel

class LazyInitMeta(type):
    _initialized = False

    def __call__(cls, *args, **kwargs):
        if not cls._initialized:
            cls.model_rebuild() # lazy init here
            cls._initialized = True
        return super().__call__(*args, **kwargs)

class CombinedMeta(LazyInitMeta, BaseModel.__class__):
    pass
```
Include it in generated package:
```toml
# pyproject.toml
...
files_to_include = [".../meta.py"]
```
Define `LazyLoadingPlugin` in `lazy_loading_plugin.py`:
```py
import ast

from ariadne_codegen.plugins.base import Plugin
from graphql import GraphQLInputObjectType


class LazyLoadingPlugin(Plugin):
    def generate_inputs_module(self, module: ast.Module) -> ast.Module:
        # adding "from .meta import CombinedMeta" to input_types.py
        module.body.insert(
            0,
            ast.ImportFrom(
                module="meta", names=[ast.alias(name="CombinedMeta")], level=1
            ),
        )

        # removing model_rebuild calls
        def _is_not_model_rebuild_call(stmt: ast.stmt) -> bool:
            return not (
                isinstance(stmt, ast.Expr)
                and isinstance(stmt.value, ast.Call)
                and isinstance(stmt.value.func, ast.Attribute)
                and stmt.value.func.attr == "model_rebuild"
            )

        module.body = list(filter(_is_not_model_rebuild_call, module.body))

        return module

    def generate_input_class(
        self, class_def: ast.ClassDef, input_type: GraphQLInputObjectType
    ) -> ast.ClassDef:
        # adding "metaclass=CombinedMeta" to every input class
        class_def.keywords.append(
            ast.keyword(arg="metaclass", value=ast.Name(id="CombinedMeta"))
        )

        return class_def

```
Add created plugin to configuration:
```toml
# pyproject.toml
...
plugins = ["....lazy_loading_plugin.LazyLoadingPlugin"]
```

**Comment by derek-upfeat** [comment]
> plugins = ["....lazy_loading_plugin.LazyLoadingPlugin"]

btw the plugin seems incorrect and i cannot figure out the correct path, in a python 3.11.2 virtual env

**Comment by mat-sop** [comment]
> Sounds great, though I don't know if we can delay those calls, would that be causing any other issues?

We're adding `model_rebuild` calls (`update_forward_refs` in `pydantic<2`) for recursive models which might be generated, depending on a specific schema. I haven't yet looked much into this, so I'm not yet sure if your solution is valid.

I found this [issue](https://github.com/pydantic/pydantic/issues/7588) on pydantic's gh, which suggests to me that there could be no documented change of behavior between `update_forward_refs` and `model_rebuild`, and we need check if those calls are still needed.

> btw the plugin seems incorrect and i cannot figure out the correct path, in a python 3.11.2 virtual env
> 
> ```shell
> ❯ tree
> .
> └── upservices
>     ├── __init__.py
>     ├── codegen.toml
>     ├── gql
>     │   └── prompts.gql
>     ├── lazy_loading_plugin.py
>     └── meta.py
>
> ```

`"....lazy_loading_plugin.LazyLoadingPlugin"` was just an example, you need to place `lazy_loading_plugin.py` in a module which is possible to import from.

In other words, this command cannot raise ImportError: `python -c "from {module_name}.lazy_loading_plugin import LazyLoadingPlugin"`, and then place `"{module_name}.lazy_loading_plugin.LazyLoadingPlugin"` into config.

**Comment by mat-sop** [comment]
Closing because in #241, I completely removed `model_rebuild` calls.

**Comment by gasbasd** [comment]
> Hey, I haven't looked in-depth into the problem yet, but for now, you can reasonably easily automate your solution by using plugin hooks, e.g.
> 
> Separate `LazyInitMeta` and `CombinedMeta` to `meta.py`:
> 
> from .base_model import BaseModel
> 
> class LazyInitMeta(type):
>     _initialized = False
> 
>     def __call__(cls, *args, **kwargs):
>         if not cls._initialized:
>             cls.model_rebuild() # lazy init here
>             cls._initialized = True
>         return super().__call__(*args, **kwargs)
> 
> class CombinedMeta(LazyInitMeta, BaseModel.__class__):
>     pass
> Include it in generated package:
> 
> # pyproject.toml
> ...
> files_to_include = [".../meta.py"]
> Define `LazyLoadingPlugin` in `lazy_loading_plugin.py`:
> 
> import ast
> 
> from ariadne_codegen.plugins.base import Plugin
> from graphql import GraphQLInputObjectType
> 
> 
> class LazyLoadingPlugin(Plugin):
>     def generate_inputs_module(self, module: ast.Module) -> ast.Module:
>         # adding "from .meta import CombinedMeta" to input_types.py
>         module.body.insert(
>             0,
>             ast.ImportFrom(
>                 module="meta", names=[ast.alias(name="CombinedMeta")], level=1
>             ),
>         )
> 
>         # removing model_rebuild calls
>         def _is_not_model_rebuild_call(stmt: ast.stmt) -> bool:
>             return not (
>                 isinstance(stmt, ast.Expr)
>                 and isinstance(stmt.value, ast.Call)
>                 and isinstance(stmt.value.func, ast.Attribute)
>                 and stmt.value.func.attr == "model_rebuild"
>             )
> 
>         module.body = list(filter(_is_not_model_rebuild_call, module.body))
> 
>         return module
> 
>     def generate_input_class(
>         self, class_def: ast.ClassDef, input_type: GraphQLInputObjectType
>     ) -> ast.ClassDef:
>         # adding "metaclass=CombinedMeta" to every input class
>         class_def.keywords.append(
>             ast.keyword(arg="metaclass", value=ast.Name(id="CombinedMeta"))
>         )
> 
>         return class_def
> Add created plugin to configuration:
> 
> # pyproject.toml
> ...
> plugins = ["....lazy_loading_plugin.LazyLoadingPlugin"]

hi @mat-sop i'm still using this plugin because without it my app consume 1GB memory, using it the memory consumption drop to 100 MB only. the only issue is that now pylance won't recognize imput classes params and complain with `No parameter named xxx`, any hint how to solve it? thank you
